spec/contracts/work_packages/base_contract_spec.rb

```diff
@@ -368,45 +378,217 @@
 
     context "when it's greater than work" do
+      let(:estimated_hours) { 5.0 }
       let(:remaining_hours) { 6.0 }
+      let(:done_ratio) { 0 }
 
       include_examples "contract is invalid", remaining_hours: :cant_exceed_work
     end
 
-    context "when unset" do
+    context "when only one being set" do
+      let(:estimated_hours) { nil }
+      let(:remaining_hours) { 1.5 }
+      let(:done_ratio) { nil }
+
+      include_examples "contract is valid"
+    end
+
+    context "when unset while work and % complete are set" do
+      let(:estimated_hours) { 3 }
+      let(:remaining_hours) { nil }
+      let(:done_ratio) { 50 }
+
+      include_examples "contract is invalid", remaining_hours: :must_be_set_when_work_and_percent_complete_are_set
+    end
+
+    context "when unset while work is set and % complete is set to an invalid value" do
+      let(:estimated_hours) { 3 }
+      let(:remaining_hours) { nil }
+      let(:done_ratio) { -1 }
+
+      include_examples "contract is invalid", estimated_hours: nil,
+                                              remaining_hours: nil,
+                                              done_ratio: :inclusion
+    end
+
+    context "when unset while % complete is set and work is set to a negative value" do
+      let(:estimated_hours) { -3 }
       let(:remaining_hours) { nil }
+      let(:done_ratio) { 50 }
 
-      include_examples "contract is invalid", remaining_hours: :must_be_set_when_work_is_set
+      include_examples "contract is invalid", estimated_hours: :greater_than_or_equal_to,
+                                              remaining_hours: nil,
+                                              done_ratio: nil
     end
   end
 
-  describe "work and remaining work" do
-    context "when changing both and remaining work exceeds work" do
+  describe "percent complete" do
+    context "when inferred by status",
+            with_settings: { work_package_done_ratio: "status" } do
+      it_behaves_like "invalid if changed", :done_ratio
+    end
+
+    context "when in work-based progress calculation mode",
+            with_settings: { work_package_done_ratio: "field" } do
       before do
-        work_package.estimated_hours = 5.0
-        work_package.remaining_hours = 6.0
+        work_package.estimated_hours = estimated_hours
+        work_package.remaining_hours = remaining_hours
+        work_package.done_ratio = done_ratio
+      end
+
+      # TODO: get better error message
+      context "when less than 0" do
+        let(:estimated_hours) { nil }
+        let(:remaining_hours) { nil }
+        let(:done_ratio) { -1 }
+
+        it_behaves_like "contract is invalid", done_ratio: :inclusion
+      end
+
+      # TODO: get better error message
+      context "when less more than 100" do
+        let(:estimated_hours) { nil }
+        let(:remaining_hours) { nil }
+        let(:done_ratio) { 101 }
+
+        it_behaves_like "contract is invalid", done_ratio: :inclusion
       end
```

Comment:
Still valid TODO comments?

Reply:
Good catch! No it's ok now. The error message is "must be between 0 and 100" whereas it was "is not included in the list" before.